### ISSUE TITLE
Negative fidelity score for SubgraphX on ba_2motifs

### ISSUE THREAD

**Issue by ShichangZh** [divelab/DIG#71]
Hello,

I run the SubgraphX model on the ba_2motifs datasets for the first 10 graphs in the dataset, and I got negative fidelity scores as the following. May I ask why this is the case and how can we reproduce the paper result on ba_2motifs? Is it because of different hyperparameters? Thanks.

```
Fidelity: -0.1811
 Infidelity: -0.1891
Sparsity: 0.8364
``` 

I trained a 3-layer GCN according to the code in `benchmarks/xgraph/train_gnns.py`, and it has accuracy 1 on the first 10 graphs I tried to explain and accuracy 0.9780 on all 1000 graphs in ba_2motifs.

I was mostly following the code in `examples/xgraph/subgraphx.ipynb` 

I constructed the explainer with the following code

```
explainer = SubgraphX(model, num_classes=2, device=device, explain_graph=True, rollout=10,
                      reward_method='mc_l_shapley', verbose=True, subgraph_building_method='zero_filling')
```
Then running the following piece of code gives the output above

```
max_nodes = 5
for i, data in enumerate(dataset[:10]):
    print(f'explain graph {i}')
    data.to(device)

    logits = model(data.x, data.edge_index)
    prediction = logits.argmax(-1).item()

    _, explanation_results, related_preds = \
        explainer(data.x, data.edge_index, max_nodes=max_nodes)

    explanation_results = explanation_results[prediction]
    explanation_results = explainer.read_from_MCTSInfo_list(explanation_results)
    result = find_closest_node_result(explanation_results, max_nodes=max_nodes)

    x_collector.collect_data(result.coalition, related_preds, label=prediction)

print(f'Fidelity: {x_collector.fidelity:.4f}\n',
      f'Infidelity: {x_collector.fidelity_inv:.4f}\n'
      f'Sparsity: {x_collector.sparsity:.4f}')
```
I also checked what's inside the X_collector, and it is the following.

```
{'_XCollector__related_preds': {'zero': [None,
   None,
   None,
   None,
   None,
   None,
   None,
   None,
   None,
   None,
   None],
  'masked': [0.9999933242797852,
   0.9999921321868896,
   0.9999922513961792,
   0.9999921321868896,
   0.9999920129776001,
   0.9999920129776001,
   0.9999923706054688,
   0.9999916553497314,
   0.9999920129776001,
   0.9999921321868896,
   0.9999921321868896],
  'maskout': [0.9999855756759644,
   0.9997314810752869,
   0.9989043474197388,
   0.9999159574508667,
   0.9341580867767334,
   0.9997014403343201,
   0.9985501170158386,
   0.9939769506454468,
   0.9990857839584351,
   0.9985829591751099,
   0.9894156455993652],
  'origin': [0.8211103081703186,
   0.8211103677749634,
   0.6541653275489807,
   0.9012890458106995,
   0.5664480924606323,
   0.6239489912986755,
   0.8487768769264221,
   0.9742612838745117,
   0.9902985692024231,
   0.8288707733154297,
   0.8895573019981384],
  'sparsity': [0.8, 0.8, 0.84, 0.84, 0.84, 0.84, 0.8, 0.92, 0.84, 0.84, 0.84]},
 '_XCollector__targets': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 'masks': [[0, 13, 15, 18, 19],
  [0, 20, 22, 23, 24],
  [0, 10, 18, 19],
  [0, 1, 14, 15],
  [21, 22, 23, 24],
  [0, 1, 5, 19],
  [0, 20, 21, 23, 24],
  [1, 15],
  [21, 22, 23, 24],
  [1, 9, 13, 14],
  [2, 7, 12, 13]],
 '_XCollector__sparsity': None,
 '_XCollector__fidelity': -0.18110649287700653,
 '_XCollector__fidelity_inv': -0.18909794092178345,
 '_XCollector__score': None}
```

**Comment by Oceanusity** [comment]
Thank you for your issue, please refer to the [repo](https://github.com/divelab/DIG/tree/main/dig/xgraph/SubgraphX) to reproduce the result of SubgraphX, and you can find hyper-parameter settings there. 

Since the `ba_2motifs` is a synthetic graph, it's quite simple and the trained model tends not to be robust. From the result you provided, it seems like this model will have a tendency to the prediction class when the input has some blocking.

**Comment by ShichangZh** [comment]
Thank you for the quick response. I will try to reproduce the SubgraphX result using the repo you referred. Nevertheless, is it all right to have negative fidelity scores? In the previous issue #47, you mentioned that the fidelity score shouldn't be negative. Also, seems like in Figure 5 in the paper, all model results on all datasets including ba_2motifs are positive.

**Comment by Oceanusity** [comment]
Yeah, it's strange to have a negative fidelity. Therefore, I recommend you to refer to the repo for details. 

One more thing, I just notice that there are several sparsity scores are more than 0.8. Since the graph in `ba_2motifs` have 25 nodes with a 5-node motif, the sparsity should be less than 0.8 to capture a complete motif. 
Welcome to post issues if you have further questions.
